File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/model/OfflineHrData.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model

import com.polar.androidcommunications.api.ble.exceptions.PmdDataParseException
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.PmdDataFrame
import com.polar.androidcommunications.common.ble.TypeUtils.convertUnsignedByteToInt

/**
 * Offline hr data
 */
internal class OfflineHrData {

    data class OfflineHrSample internal constructor(
        val hr: Int,
        val ppgQuality: Int,
        val correctedHr: Int
    )

    val hrSamples: MutableList<OfflineHrSample> = mutableListOf()

    companion object {
        fun parseDataFromDataFrame(frame: PmdDataFrame): OfflineHrData {
            return if (frame.isCompressedFrame) {
                throw PmdDataParseException(
                    "Offline HR compressed frame type ${frame.frameType} is not supported"
                )
            } else {
                when (frame.frameType) {
                    PmdDataFrame.PmdDataFrameType.TYPE_0 -> dataFromType0(frame)
                    PmdDataFrame.PmdDataFrameType.TYPE_1 -> dataFromType1(frame)
                    else -> throw PmdDataParseException(
                        "Offline HR raw frame type ${frame.frameType} is not supported"
                    )
                }
            }
        }

        private fun dataFromType0(frame: PmdDataFrame): OfflineHrData {
            // Offline HR data contains HR values — raw bytes are NOT included in exception
            // messages as they may contain sensitive health information.
            if (frame.dataContent.isEmpty()) {
                throw PmdDataParseException(
                    "Offline HR raw TYPE_0 dataContent is empty"
                )
            }
            var offlineHrData = OfflineHrData()
            var offset = 0
            while (offset < frame.dataContent.size) {
                val hr = convertUnsignedByteToInt(frame.dataContent[offset])
                offlineHrData.hrSamples.add(OfflineHrSample(hr, 0, 0))
                offset += 1
            }
            return offlineHrData
        }

        private fun dataFromType1(frame: PmdDataFrame): OfflineHrData {
            // Offline HR data contains HR values — raw bytes are NOT included in exception
            // messages as they may contain sensitive health information.
            if (frame.dataContent.isEmpty() || frame.dataContent.size % 3 != 0) {
                throw PmdDataParseException(
                    "Offline HR raw TYPE_1 dataContent size ${frame.dataContent.size} is not a " +
                    "non-zero multiple of expected sample size 3"
                )
            }
            val offlineHrData = OfflineHrData()
            var offset = 0
            while (offset < frame.dataContent.size) {
                val hr = convertUnsignedByteToInt(frame.dataContent[offset])
                offset += 1
                val ppgQual = convertUnsignedByteToInt(frame.dataContent[offset])
                offset += 1
                val correctedHr = convertUnsignedByteToInt(frame.dataContent[offset])
                offlineHrData.hrSamples.add(OfflineHrSample(hr, ppgQual, correctedHr))
                offset += 1
            }
            return offlineHrData
        }
    }
}

File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/gatt/client/pmd/model/OfflineHrData.swift
//  Copyright © 2023 Polar. All rights reserved.

import Foundation

public class OfflineHrData {
    
    struct OfflineHrSample {
        let hr: UInt8
        let ppgQuality: UInt8
        let correctedHr: UInt8
    }
    
    var samples: [OfflineHrSample]
    
    init(samples: [OfflineHrSample] = []) {
        self.samples = samples
    }
    
    static func parseDataFromDataFrame(frame: PmdDataFrame) throws -> OfflineHrData {
        if (frame.isCompressedFrame) {
            throw PmdDataParseError(message: "Compressed FrameType: \(frame.frameType) is not supported by Offline HR data parser")
        } else {
            switch (frame.frameType) {
            case PmdDataFrameType.type_0: return try dataFromRawType0(frame: frame)
            case PmdDataFrameType.type_1: return try dataFromRawType1(frame: frame)
            default: throw PmdDataParseError(message: "Raw FrameType: \(frame.frameType) is not supported by Offline HR data parser")
            }
        }
    }
    
    private static func dataFromRawType0(frame: PmdDataFrame) throws -> OfflineHrData {
        let offlineHrData = OfflineHrData()
        var offset = 0
        while (offset < frame.dataContent.count) {
            offlineHrData.samples.append(OfflineHrSample(hr: UInt8(frame.dataContent[offset]), ppgQuality: 0, correctedHr: 0))
            offset += 1
        }
        return offlineHrData
    }

    private static func dataFromRawType1(frame: PmdDataFrame) throws -> OfflineHrData {
        let sampleByteSize = 3
        guard !frame.dataContent.isEmpty && frame.dataContent.count % sampleByteSize == 0 else {
            throw PmdDataParseError(
                message: "OfflineHR raw TYPE_1 dataContent size \(frame.dataContent.count) is not a " +
                "non-zero multiple of expected sample size \(sampleByteSize)")
        }
        let offlineHrData = OfflineHrData()
        var offset = 0
        while (offset < frame.dataContent.count) {
            let hr = UInt8(frame.dataContent[offset])
            offset += 1
            let ppgQual = UInt8(frame.dataContent[offset])
            offset += 1
            let correctedHr = UInt8(frame.dataContent[offset])
            offlineHrData.samples.append(OfflineHrSample(hr: hr, ppgQuality: ppgQual, correctedHr: correctedHr))
            offset += 1
        }
        return offlineHrData
    }
}
